### PR TITLE
Docking as a Feature Mk2

### PR DESCRIPTION
**Pull request by dbr**
As described in https://github.com/imgui-rs/imgui-rs/pull/538#issuecomment-951504736 - this is just the "boring" parts of that PR which revise the build system, and `#[cfg(...)]` around the few differences in the API in the upstream `docking` branch

Meaning this basically just allows user to enable the `imgui/docking` feature, then do
```rust
imgui_ctx.io_mut().config_flags |= imgui::ConfigFlags::DOCKING_ENABLE;
```
to enable windows to be docked together

Once this is good and merged, we can start deciding which bits of the docking API to expose and how (probably very roughly [along these lines](https://github.com/dbr/imgui-rs/blob/duckfeet/imgui/src/docking.rs))

### CODE REVIEW

File: xtask/src/bindgen.rs [hunk 1]
```diff
@@ -104,7 +110,7 @@ fn generate_binding_file(
         "--no-prepend-enum-name",
         "--no-doc-comments",
         // Layout tests aren't portable (they hardcode type sizes), and for
-        // our case they just serve to sanity check rustc's implementation of
```
[the remark this answers is not in the document]
Comment (dbr) [reply xtask/src/bindgen.rs:107]: Ah it had nonbreaking-space between our/case

File: imgui-sys/build.rs [hunk 2]
```diff
@@ -27,23 +26,18 @@ fn assert_file_exists(path: &str) -> io::Result<()> {
 }
 
 fn main() -> io::Result<()> {
-    let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
-    println!(
```
[the remark this answers is not in the document]
Comment (dbr) [reply imgui-sys/build.rs:31]: Ohh I didn't realize this - wasn't clear what `cargo:THIRD_PARTY` did, I had assumed it was some legacy thing as nothing seems to break without it.. but it being used by implot/etc makes perfect sense

Should be easy enough to restore this keeping compatibility I think - pointing same var to equivelant folder (which is the same just extra directory level) should function the same, as far as I can tell

File: imgui-sys/src/lib.rs [hunk 3]
```diff
@@ -20,17 +20,29 @@
...
-
-#[cfg(feature = "wasm")]
-pub use crate::wasm_bindings::*;
-
-#[cfg(not(feature = "wasm"))]
-mod bindings;
-
-#[cfg(not(feature = "wasm"))]
-pub use crate::bindings::*;
+cfg_if::cfg_if! {
```
[the remark this answers is not in the document]
Comment (dbr) [reply imgui-sys/src/lib.rs:23]: I couldn't work out a neater way of doing this - how would you write the flattened version?

File: imgui-sys/src/bindings.rs [hunk 4]
```diff
@@ -164,7 +163,11 @@ pub struct ImVector_ImDrawChannel {
 }
 impl Default for ImVector_ImDrawChannel {
     fn default() -> Self {
-        unsafe { ::core::mem::zeroed() }
+        let mut s = ::core::mem::MaybeUninit::<Self>::uninit();
```
[the remark this answers is not in the document]
Comment (dbr) [reply imgui-sys/src/bindings.rs:166]: Does seem a little roundabout - no idea why - there's nothing obvious in the bindgen changelog between 0.56 and 0.59 which would explain why this changed

File: imgui-sys/build.rs [hunk 5]
```diff
@@ -58,21 +52,28 @@ fn main() -> io::Result<()> {
             build.define("IMGUI_ENABLE_FREETYPE", None);
             println!("cargo:DEFINE_IMGUI_ENABLE_FREETYPE=");
 
-            // imgui_freetype.cpp needs access to imgui.h
-            build.include(manifest_dir.join("third-party/imgui/"));
+            // imgui_freetype.cpp needs access to `#include "imgui.h"`
+            let manifest_dir = std::path::Path::new(env!("CARGO_MANIFEST_DIR"));
+            #[cfg(feature = "docking")]
```
[the remark this answers is not in the document]
Comment (dbr) [reply imgui-sys/build.rs:57]: That makes sense - might as well make the change now, would probably make the code a little clearer too (I don't like the `cfg(...)\ncfg(not(...))` conditionals here)

File: imgui-sys/build.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbr) [reply imgui-sys/build.rs:31]: Restored this env-var and it should, if I'm understanding right, remain compatible

**Comment by dbr** [comment]
> but I think I'd prefer the name imgui-release to imgui-master for the directory, since it's not actually taken from the master branch (it's taken from a tag off of it, presumably)

The `imgui-docking` is also not taken from the latest `docking` ref directly, but a commit on that branch which corresponds to the tagged release, so I think the current naming is clearest approach. If the default branch is renamed we can update here without a breaking change

> We need to still expose a way to access DEP_IMGUI_THIRD_PARTY

A good point, I wasn't aware of this - this is now restored and should work exactly as it did before.

> At the very least, IMO you should write down the concrete steps to take to perform the update

Done in `docs/upgrading-imgui.md` (and I'll reference this file in that thread for visibility)

**Comment by sanbox-irl** [comment]
@dbr lets do it?

**Comment by dbr** [comment]
:runner:

**Pull request merged**